File: examples/python/12_diagnosis.py
"""12 · Diagnosis — when retrieval looks weak, the Decision Report tells you why.

Real-world scenario:
    A support team is wiring up Q&A over a policy doc. A user asks
    *"how long do I have to cancel and get my money back?"* and gets
    an empty answer. The doc uses *refund* and *termination*, not
    *cancel* and *money back*, so BM25 has nothing to match.

What this demonstrates:
    - `ctx.report.diagnosis` populated on every `context()` call.
    - Layer-2 facts: `query_terms`, `zero_match_terms`, `term_stats`
      computed against the corpus vocabulary.
    - The closed `hints` registry: one bounded hint per documented
      failure shape, each citing the finding that justifies it.
    - A healthy query produces zero hints. No spam on the happy path.

Run:
    pip install redhop
    python examples/python/12_diagnosis.py
"""

import redhop


def main() -> None:
    doc = redhop.Document.from_chunks(
        [
            redhop.Chunk(
                "Refund Policy. Refunds are available within thirty days of purchase.",
                id="a",
                source="policy.md",
            ),
            redhop.Chunk(
                "Termination for convenience. Either party may terminate this agreement.",
                id="b",
                source="policy.md",
            ),
            redhop.Chunk(
                "Governing Law. This agreement is governed by the laws of California.",
                id="c",
                source="policy.md",
            ),
        ]
    )

    # ── 1. A healthy query: facts populated, no hints ──────────────────
    healthy = doc.context("refund policy thirty days")
    d = healthy.report.diagnosis
    print("Healthy query:")
    print(f"  query_terms              = {d['query_terms']}")
    print(f"  corpus_stats_available   = {d['corpus_stats_available']}")
    print(f"  zero_match_terms         = {d['zero_match_terms']}")
    print(f"  hints                    = {d['hints']}")
    print()

    # ── 2. Vocabulary-mismatch query: H2 fires, evidence cited ────────
    paraphrase = doc.context("How long do I have to cancel and get my money back?")
    d = paraphrase.report.diagnosis
    print("Vocabulary-mismatch query:")
    print(f"  query_terms              = {d['query_terms']}")
    print(f"  zero_match_terms         = {d['zero_match_terms']}")
    print(f"  empty_context            = {d['empty_context']}")
    for hint in d["hints"]:
        print(f"  hint {hint['code']!r}")
        print(f"    evidence  : {hint['evidence']}")
        print(f"    message   : {hint['message']}")
    print()

    # ── 3. The rendered Decision Report shows the same data ──────────
    # (Format conventions match the rest of the report. No "Query
    # diagnosis" section appears for healthy calls; on bad calls it
    # appears below the existing Warnings block.)
    print("Rendered report (excerpt):")
    rendered = str(paraphrase.report)
    if "Query diagnosis" in rendered:
        section = rendered[rendered.index("Query diagnosis"):]
        print(section)


if __name__ == "__main__":
    main()

File: examples/rust/examples/12_diagnosis.rs
//! 12 · Diagnosis — when retrieval looks weak, the Decision Report tells you why.
//!
//! Real-world scenario:
//!   A support team is wiring up Q&A over a policy doc. A user asks
//!   "how long do I have to cancel and get my money back?" and gets
//!   an empty answer. The doc uses *refund* and *termination*, not
//!   *cancel* and *money back*, so BM25 has nothing to match.
//!
//! What this demonstrates:
//!   - `report.diagnosis` populated on every `context()` call.
//!   - Layer-2 facts: `query_terms`, `zero_match_terms`, `term_stats`
//!     computed against the corpus vocabulary.
//!   - The closed hints registry: one bounded hint per documented
//!     failure shape, each citing the finding that justifies it.
//!   - A healthy query produces zero hints.
//!
//! Run:
//!   cargo run --example 12_diagnosis

use redhop::{Chunk, ChunkId, Document, HintCode, TokenCount};

fn corpus() -> Vec<Chunk> {
    vec![
        Chunk::new(
            ChunkId::new("a"),
            "Refund Policy. Refunds are available within thirty days of purchase.",
            "policy.md",
            TokenCount(11),
        ),
        Chunk::new(
            ChunkId::new("b"),
            "Termination for convenience. Either party may terminate this agreement.",
            "policy.md",
            TokenCount(10),
        ),
        Chunk::new(
            ChunkId::new("c"),
            "Governing Law. This agreement is governed by the laws of California.",
            "policy.md",
            TokenCount(11),
        ),
    ]
}

fn main() -> redhop::Result<()> {
    let mut doc = Document::from_chunks(corpus())?;

    // 1. Healthy query: facts populated, no hints.
    let healthy = doc.context("refund policy thirty days")?;
    let d = &healthy.report.diagnosis;
    println!("Healthy query:");
    println!("  query_terms            = {:?}", d.query_terms);
    println!("  corpus_stats_available = {}", d.corpus_stats_available);
    println!("  zero_match_terms       = {:?}", d.zero_match_terms);
    println!("  hints                  = {}", d.hints.len());
    println!();

    // 2. Vocabulary-mismatch query: VocabMismatch hint fires.
    let paraphrase = doc.context("How long do I have to cancel and get my money back?")?;
    let d = &paraphrase.report.diagnosis;
    println!("Vocabulary-mismatch query:");
    println!("  query_terms            = {:?}", d.query_terms);
    println!("  zero_match_terms       = {:?}", d.zero_match_terms);
    println!("  empty_context          = {}", d.empty_context);
    for hint in &d.hints {
        let label = match hint.code {
            HintCode::EmptyContext => "empty_context",
            HintCode::VocabMismatch => "vocab_mismatch",
            HintCode::LowConfidence => "low_confidence",
            HintCode::LowDiscriminationQuery => "low_discrimination_query",
            HintCode::UnderdeterminedQuery => "underdetermined_query",
            _ => "unknown",
        };
        println!("  hint {:?}", label);
        println!("    evidence  : {}", hint.evidence);
        println!("    message   : {}", hint.message);
    }
    println!();

    // 3. Rendered report (the same data, human-readable).
    let rendered = paraphrase.report.render(None);
    if let Some(idx) = rendered.find("Query diagnosis") {
        println!("Rendered report (excerpt):");
        println!("{}", &rendered[idx..]);
    }

    Ok(())
}
